pkg/controller/test-images-distributor/test_images_distributor.go

```diff
@@ -0,0 +1,364 @@
+
+func decodeRequest(req reconcile.Request) (string, types.NamespacedName, error) {
+	clusterAndNamespace := strings.Split(req.Namespace, "_")
+	if n := len(clusterAndNamespace); n != 2 {
+		return "", types.NamespacedName{}, fmt.Errorf("didn't get two but %d segments when trying to extract cluster and namespace", n)
+	}
+	return clusterAndNamespace[0], types.NamespacedName{Namespace: clusterAndNamespace[1], Name: req.Name}, nil
+}
+
+type reconciler struct {
+	ctx                      context.Context
+	log                      *logrus.Entry
+	registryClient           ctrlruntimeclient.Client
+	buildClusterClients      map[string]ctrlruntimeclient.Client
+	pullSecretGetter         func() []byte
+	successfulImportsCounter *prometheus.CounterVec
+	failedImportsCounter     *prometheus.CounterVec
+	dryRun                   bool
+}
+
+func (r *reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithField("request", req.String())
+	log.Info("Starting reconciliation")
+	err := r.reconcile(req, log)
+	if err != nil {
+		log.WithError(err).Error("Reconciliation failed")
+	} else {
+		log.Info("Finished reconciliation")
+	}
+	return reconcile.Result{}, controllerutil.SwallowIfTerminal(err)
+}
+
+func (r *reconciler) reconcile(req reconcile.Request, log *logrus.Entry) error {
+	cluster, decoded, err := decodeRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to decode request %s: %w", req, err)
+	}
+	// Propagate the cluster field back up
+	*log = *log.WithField("cluster", cluster)
+
+	// Fail asap if we cannot reconcile this
+	client, ok := r.buildClusterClients[cluster]
+	if !ok {
+		return controllerutil.TerminalError(fmt.Errorf("no client for cluster %s available", cluster))
+	}
+
+	sourceImageStreamTag := &imagev1.ImageStreamTag{}
+	if err := r.registryClient.Get(r.ctx, decoded, sourceImageStreamTag); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("Source imageStreamTag not found")
+			return nil
+		}
+		return fmt.Errorf("failed to get imageStreamTag %s from registry cluster: %w", decoded.String(), err)
+	}
+
+	isCurrent, err := r.isImageStreamTagCurrent(decoded, client, sourceImageStreamTag)
+	if err != nil {
+		return fmt.Errorf("failed to check if imageStreamTag %s on cluster %s is current: %w", decoded.String(), cluster, err)
+	}
+	if isCurrent {
+		log.Debug("ImageStreamTag is current")
+		return nil
+	}
+
+	if err := client.Get(r.ctx, types.NamespacedName{Name: decoded.Namespace}, &corev1.Namespace{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check if namespace %s exists: %w", decoded.Namespace, err)
+		}
+		if err := client.Create(r.ctx, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: decoded.Namespace}}); err != nil && !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create namespace %s: %w", decoded.Namespace, err)
+		}
+	}
+
+	if err := r.ensureImagePullSecret(decoded.Namespace, client); err != nil {
```

Comment:
Not sure if its the right thing to have this here. We also have that secret mirroring controller somewhere which could do it as well. The thought was that this easens debugging and testing, because we don't have our business logic scattered around in being partly here and partly done by configuring a generic controller to do the subset of our businesslogic its able to do.

Reply:
Yeah I think this is fine